### ISSUE TITLE
Handle screen size changes while calling getch

### ISSUE THREAD

**Issue by mcormier** [ruby/curses#9]
Attached is a simple curses program that displays some text and calls getch. It also sets up a trap on SIGWINCH so that it can detect window size changes.

``` ruby
require 'curses'
include Curses

@listeners = []

def size_changed
  # TODO -- notify all size changed listeners
  @listeners.each { |listener|
    #listener.screen_size_changed
 }
end


begin
  Signal.trap('SIGWINCH', size_changed )

  init_screen
  Curses.raw
  clear
  curs_set(0)
  noecho
  cbreak
  start_color

  Curses.stdscr.setpos(0, 0)
  Curses.stdscr.addstr("Press any key ")
  Curses.stdscr.getch

rescue SystemExit, Interrupt
  # Empty Catch block so ruby doesn't puke out
  # a stack trace when CTRL-C is used
ensure
  Curses.close_screen
end
```

The issue is that the call to an array method causes a NoMethodError when the size_changed method gets called.

To reproduce:
1. Run the sample program
2. Resize the terminal window 

![untitled](https://cloud.githubusercontent.com/assets/39216/5063710/a1d4f4ce-6dc7-11e4-85cc-1a22e107d0af.png)

I also found this post:
http://masanjin.net/blog/ruby-ncurses-and-thread-blocking

That talks about some implementations of curses blocking threads on getch.  

I took a look at the ncurses-ruby gem and found some interesting code in C, entitled rbncurshelper_nonblocking_wgetch

https://github.com/eclubb/ncurses-ruby/blob/master/ext/ncurses/ncurses_wrap.c#L792

Basically,  I'm writing a GUI library on top of curses and would like to handle window resizes by notifying all objects to redraw themselves.  Was wondering.
1. Are there limitations to what you can do when handling a signal?
2. Is getch blocking?
3. Is there a better way to handle this then catching NoMethodError's like below

``` ruby
 def get_ch_handle_signals
      got_input = false
      while !got_input
        begin
          c = getch
          got_input = true
        rescue NoMethodError
          # TODO -- SIGWINCH could have occurred -- reposition..
          c = ''
        end
      end

      c
    end
```

**Comment by shugo** [comment]
@mcormier You need to refresh the screen as follows:

```
Thread.new do
  sleep 0.1
  Curses.stdscr.setpos(0, 0)
  Curses.stdscr.addstr('library is GOOD')
  Curses.stdscr.refresh
end
```

You don't need to handle SIGWINCH on you own because getch returns
Curses::KEY_RESIZE when SIGWINCH is detected.

**Comment by shugo** [comment]
@mcormier 

> ```
> def size_changed
>   # TODO -- notify all size changed listeners
>   @listeners.each { |listener|
>     #listener.screen_size_changed
>  }
> end
> 
> 
> begin
>   Signal.trap('SIGWINCH', size_changed )
> ```

The code doesn't work because size_changed is called at the time when Signal.trap is called.

Try the following code:

```ruby
def size_changed(sig)
  # TODO -- notify all size changed listeners
  @listeners.each { |listener|
    #listener.screen_size_changed
  }
end


begin
  Signal.trap('SIGWINCH', method(:size_changed))
```

Or more simply:

```ruby
begin
  Signal.trap('SIGWINCH') do
    ...
  end
```
